index_votes.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import os
+import textwrap
+from psycopg2.errors import ForeignKeyViolation
+import requests
+from utils import (
+    is_archive_node,
+    PollingProcess,
+    events_to_process,
+    new_events_to_process,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def fetch_votes_by_proposal(height, proposal_id):
+    if is_archive_node():
+        headers = {"x-cosmos-block-height": str(height)}
+    else:
+        headers = None
+    # Currently EventVote only contains proposal id
+    # Eventually EventVote may contain proposal id and voter address
+    # At which point we could get the vote with this endpoint:
+    # /cosmos/group/v1/vote_by_proposal_voter/{proposal_id}/{voter}
+    # Ref: https://github.com/regen-network/indexer/pull/38#discussion_r1310958235
+    resp = requests.get(
+        f"{os.environ['REGEN_API']}/cosmos/group/v1/votes_by_proposal/{proposal_id}",
+        headers=headers,
+    )
+    resp.raise_for_status()
+    return resp.json()["votes"]
+
+
+def gen_records(cur, query):
+    cur.execute(query)
+    for record in cur:
+        yield record
+
+
+def _index_votes(pg_conn, _client, _chain_num):
+    with pg_conn.cursor() as cur:
+        all_chain_nums = [
+            record[0] for record in gen_records(cur, "select num from chain;")
+        ]
+        max_block_heights = {
+            chain_num: max_block_height
+            for chain_num, max_block_height in gen_records(
+                cur,
+                "select chain_num, MAX(block_height) from votes group by chain_num;",
+            )
+        }
+        logger.debug(f"{all_chain_nums=}")
+        logger.debug(f"{max_block_heights=}")
+        for chain_num in all_chain_nums:
+            if chain_num not in max_block_heights.keys():
+                max_block_heights[chain_num] = 0
+        logger.debug(f"{max_block_heights=}")
+        for chain_num, max_block_height in max_block_heights.items():
+            logger.debug(f"{chain_num=} {max_block_height=}")
```

Comment:
Also noticing we no longer make use of the `_chain_num` parameter with this addition.

Reply:
> Maybe this is ok for now but this should either be extracted to reuse in other processes or we should rethink how we are running processes, i.e. whether we want to run separate processes for separate chains.

we can figure out a way to extract it. maybe we can do that as a part of refactoring indexing credit class issuers to use the same method.

Reply:
> Not sure if we should address this issue (#33) separately. Any reason for adding it to this pull request?

it’s not really added to this PR. it’s just that this PR will index votes in a forward looking way. if we want to index multiple chains in the same processes, this process will still work for that. (as do all the other processes we’ve written to date).

Reply:
> it’s not really added to this PR. it’s just that this PR will index votes in a forward looking way

It's not included in the other processes, e.g. `index_proposals.py`, `index_class_issuers`, etc. We are adding a loop here that does not exist in the other processes, and which seems more relevant to #33 than #22. This pull request is already addressing two issues. It would be nice to focus on one issue at a time.

> if we want to index multiple chains in the same processes, this process will still work for that

As mentioned in the review comment, I'm not sure one process for all chains is the best approach. It might be easier to run multiple instances of the process, one for each chain. We have a single environment variable for a a regen node. It is  diverging from the current architecture and its not required for indexing votes as far as I can tell.

> this is an unused parameter for all the indexing processes so far i.e.

Why was it added in the first place? Is there a future use case for it?

Reply:
> Why was it added in the first place? Is there a future use case for it?

I honestly forgot why I put it there...

> Maybe it could be useful if we want to run multiple instances of the process for each chain?

But yes indeed, it can be useful for this.

> Adding the loop on max_block_heights.items() is needed for the new new_events_to_process logic. But the previous loop on all_chain_nums could potentially be avoided if we use the _chain_num param instead and distribute this logic into different processes.

Yes, I think a good follow up task will be to simplify this as you've described here:

https://github.com/regen-network/indexer/issues/44

This is also something that @ryanchristo and I discussed last week, so good that we are all in agreement it's a good thing to do.